docs/specs/authentication-flow-prioritized-identity.md

```diff
@@ -0,0 +1,268 @@
+```
+
+`identification: email` was chosen according to the input. This option has a `priority` value of `0`, which is the default value because it is not given.
+
+Looking at other existing options, we found there is another option `identification: oauth`. And the `priority` value is `1` for this option. `1` > `0`, so this option has a higher priority value than the chosen identification method.
+
+Looking at the existing identities of Alice, she has a Google oauth identity, which can be used to pass the `identification: oauth`. As a result, an error will be returned by the authentication flow API:
+
+```jsonc
+{
+  "name": "Invalid",
+  "reason": "UnpreferredIdentity",
+  "message": "please use another identification method",
+  "code": 400,
+  "info": {
+    "PreferredIdentitifications": [
+      {
+        "identification": "oauth",
+        "provider_type": "google",
+        "alias": "google"
+      }
+    ]
+  }
+}
+```
+
+When Alice saw this error, she knows she should use `oauth` identification method to login. So passing in a new input:
+
+```jsonc
+{
+  "identification": "oauth",
+  "alias": "google",
+  "redirect_uri": "http://example.com/sso/oauth2/callback/google"
+}
+```
+
+This time, `identification: oauth` was chosen according to the input. This option has a `priority` value of `1`.
+
+Looking at other existing options, there is no other options with priority higher than `1`, therefore she is able to pass the step and continue the login flow.
+
+## Behaviors
+
+This section specifies the behavior of `priority` under different cases.
+
+### When there are multiple levels of priority
+
+Assume we have the following config:
+
+```yaml
+authentication_flow:
+  login_flows:
+    - name: default
+      steps:
+        - name: identify
+          type: identify
+          one_of:
+            - identification: oauth
+              priority: 3
+            - identification: phone
+              priority: 2
+            - identification: email
+              priority: 1
+            - identification: username
+              priority: 0
+```
+
+In this example, we have three levels of `priority`. `0`, `1` and `2`.
+
+When the following input was passed in:
+
+```json
+{
+  "identification": "email",
+  "login_id": "alice@example.com"
+}
+```
+
+`identification: email` will be selected as the identification method, which has `priority` of `1`. `phone` and `oauth` which has priority higher than the current selected identification method `email` will be included in the `PreferredIdentitifications` array, while `username` will be excluded.
+
+```jsonc
+{
+  "name": "Invalid",
+  "reason": "UnpreferredIdentity",
+  "message": "please use another identification method",
+  "code": 400,
+  "info": {
+    "PreferredIdentitifications": [
+      {
+        "identification": "oauth",
+        "provider_type": "google",
+        "alias": "google"
+      },
+      {
+        "identification": "phone"
```

Comment:
Why show both oauth/google and phone here? Isn't that we should show the "ONE" identification method that this users could use?

Reply:
My consideration is, if we only return a list of options which the user has, it seems we exposed more imformation about what identities the user has.
However, after rethink it seems we already exposed the imformation when we have different behavior on user has / doesn't have the corresponding identity, so looks like it is not very critical.
Let me change it to "Only return the options that user can choose", which is more suitable for the usecase.